### DISCUSSION TITLE
Return deeply nested Java object to Javascript without access to source code

### DISCUSSION THREAD

**Discussion by dangmai** [konsoletyper/teavm#958] [Q&A] [answered]
Hello there,

First of all thank you for making TeaVM. It's an extremely ambitious project, and I'm very pleasantly surprised at how far I can get out of the box with it (I'm doing some comparisons between various toolings - TeaVM, J2CL, Bytecoder, etc.), and TeaVM's the only one getting me fairly close to what I'm trying to achieve. Kudos to that 💯 

A bit of context about my question: I am dealing with a closed source JAR whose source code I do not have access to for my open source project [Prettier Apex](https://github.com/dangmai/prettier-plugin-apex). I have permission to use that JAR from the company generating it (Salesforce), but I do not have access to its source code. The library basically parses source code into an Abstract Syntax Tree structure.

I have managed to transpile this JAR into an ES2025 module which can be imported by my NodeJS script, after adding in a bunch of fake classlibs + re-implement a few methods. I've annotated a [method](https://github.com/dangmai/prettier-plugin-apex/blob/13531f1a2904419da87c5170161091b0fe33d5bd/packages/apex-ast-serializer/parser/src/main/java/net/dangmai/serializer/DirectApex.java#L19) as `@JSExport`, and I can now call it from NodeJS and getting back a tree structure. However, the nodes under the root all have mangled names. From what I understand, I have to annotate all getters for all node classes with `@JSProperty` in order for them to have correct names when returned to JS. But since I don't have access to the source code, is there something else that I can do to make this work?

I've looked into the Metaprogramming guide, but it doesn't look like I can add annotation to a ReflectField either.

I can also make the child node's fields less mangled in JS by setting `obfuscated = False`, but I think that those names are implementation details that I shouldn't rely on. Also, even when exporting non-obfuscated code, the field names still have number suffix to them, which would make reading the correct fields more difficult.

Thanks again!

**Comment by konsoletyper** [comment, marked as answer]
Hi. I think TeaVM itself can't help you. You can try following solutions:

1. Edit JAR file's classes using some Java bytecode manipulation library (like ASM) to add missing annotations
2. Instead of returning library objects from your module, write your own wrappers which would just delegate calls to library classes and annotate your wrappers with `@JSExport`.

**Reply by dangmai** [reply]
Thank you for answering my question. I was already playing around with adding annotations via bytecode manipulation, and just wanna check if I was missing something that might be easier.
I appreciate it, you rock! 🥇

**Reply by dangmai** [reply]
Adding annotations via bytecode manipulation did work for the classes inside the JAR. However, the tree structure also uses certain TeaVM-provided classes (e.g. `ArrayList`), which do not have the annotations in them. What's my best course of action for these classes? I guess I could copy paste the TArrayList source code then add my own annotations in it, but that doesn't feel very maintainable for the future.

**Reply by konsoletyper** [reply]
Sorry, I'm afraid I don't have any solution for this
